cpp/src/dbscan/dbscan.cuh

```diff
@@ -207,5 +211,223 @@ void dbscanFitImpl(const raft::handle_t& handle,
+                   Index_* n_rows_ptr,
+                   Index_ n_cols,
+                   const T* eps_ptr,
+                   const Index_* min_pts_ptr,
+                   raft::distance::DistanceType metric,
+                   Index_* labels,
+                   Index_* core_sample_indices,
+                   size_t max_mbytes_per_dispatch,
+                   cudaStream_t stream,
+                   int verbosity,
+                   void* custom_workspace        = nullptr,
+                   size_t* custom_workspace_size = nullptr)
+{
+  raft::common::nvtx::range fun_scope("ML::Dbscan::Fit");
+  ML::Logger::get().setLevel(verbosity);
+  /// NOTE: for algo_vd, 0 (naive) and 2 (precomputed) are not supported.
+  ///       for algo_adj, 0 (naive) is not supported.
+  int algo_vd  = 1;
+  int algo_adj = 1;
+  int algo_ccl = 2;
+
+  Index_ accum_rows = thrust::reduce(thrust::host, n_rows_ptr, n_rows_ptr + n_groups);
+  ASSERT(accum_rows > 0, "No rows in the input array. DBSCAN cannot be fitted!");
+
+  // Estimate available memory per dispatch
+  // Note: we can't rely on the reported free memory.
+  if (max_mbytes_per_dispatch == 0) {
+    // Query memory information to get the total memory on the device
+    size_t free_memory, total_memory;
+    RAFT_CUDA_TRY(cudaMemGetInfo(&free_memory, &total_memory));
+
+    size_t dataset_memory = ((size_t)accum_rows * (size_t)n_cols * sizeof(T));
+
+    // The estimate is: 80% * total - dataset
+    max_mbytes_per_dispatch = (80 * total_memory / 100 - dataset_memory) / 1e6;
+
+    CUML_LOG_DEBUG("Dataset memory: %ld MB", (unsigned long long)(dataset_memory / 1e6));
+
+    CUML_LOG_DEBUG("Estimated available memory: %ld / %ld MB",
+                   (unsigned long long)max_mbytes_per_dispatch,
+                   (unsigned long long)(total_memory / 1e6));
+  }
+
+  CUML_LOG_DEBUG("Scheduling input groups");
+  std::vector<std::vector<Index_>> grouped_row_ids;
+  mgrp_dbscan_scheduler<Index_>(
+    max_mbytes_per_dispatch, n_groups, n_rows_ptr, n_rows_ptr, grouped_row_ids);
+  CUML_LOG_DEBUG("Divide input into %lu groups", grouped_row_ids.size());
+  if (verbosity >= CUML_LEVEL_DEBUG) {
+    std::cout << "row_groups: [ ";
+    for (auto group : grouped_row_ids) {
+      std::cout << "[";
+      for (auto it : group) {
+        std::cout << n_rows_ptr[it] << " ";
+      }
+      std::cout << "];";
+    }
+    std::cout << std::endl;
+  }
+
+  std::vector<Index_> pfx_rows(n_groups, 0);
+  thrust::exclusive_scan(thrust::host, n_rows_ptr, n_rows_ptr + n_groups, pfx_rows.data());
+
+  // Get the maximum of workspace size
+  size_t max_workspace_size = 0;
+  for (size_t i = 0; i < grouped_row_ids.size(); ++i) {
+    std::vector<Index_> dispatch_group = grouped_row_ids[i];
+    Index_ dispatch_n_groups           = dispatch_group.size();
+    Index_ first_row_id                = dispatch_group[0];
+    Index_* dispatch_n_rows_ptr        = n_rows_ptr + first_row_id;
+    if (dispatch_n_groups != 1) {
```

Comment:
That's because when `dispatch_n_groups == 1`, it falls back to origin DBSCAN straightforwardly, which undertakes the responsibility of calculation of workspace size and it slightly differs from other cases.